x-pack/elastic-agent/pkg/core/plugin/service/app.go

```diff
@@ -0,0 +1,330 @@
+func (a *Application) SetState(status state.Status, msg string) {
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+	a.setState(status, msg)
+}
+
+// Start starts the application with a specified config.
+func (a *Application) Start(ctx context.Context, t app.Taggable, cfg map[string]interface{}) (err error) {
+	defer func() {
+		if err != nil {
+			// inject App metadata
+			err = errors.New(err, errors.M(errors.MetaKeyAppName, a.name), errors.M(errors.MetaKeyAppName, a.id))
+		}
+	}()
+
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+
+	cfgStr, err := yaml.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	// already started
+	if a.srvState != nil {
+		a.setState(state.Starting, "Starting")
+		a.srvState.SetStatus(proto.StateObserved_STARTING, a.state.Message)
+		a.srvState.UpdateConfig(string(cfgStr))
+	} else {
+		a.setState(state.Starting, "Starting")
+		a.srvState, err = a.srv.Register(a, string(cfgStr))
+		if err != nil {
+			return err
+		}
+	}
+
+	defer func() {
+		if err != nil {
+			if a.srvState != nil {
+				a.srvState.Destroy()
+				a.srvState = nil
+			}
+		}
+	}()
+
+	if err := a.monitor.Prepare(a.name, a.pipelineID, a.uid, a.gid); err != nil {
+		return err
+	}
+
+	if a.limiter != nil {
+		a.limiter.Add()
+	}
+
+	// start the credentials listener for the service
+	if err := a.startCredsListener(); err != nil {
+		return err
+	}
+
+	// allow the service manager to ensure that the application is started, currently this does not start/stop
+	// the actual service in the system service manager
+
+	return nil
+}
+
+// Configure configures the application with the passed configuration.
+func (a *Application) Configure(_ context.Context, config map[string]interface{}) (err error) {
+	defer func() {
+		if err != nil {
+			// inject App metadata
+			err = errors.New(err, errors.M(errors.MetaKeyAppName, a.name), errors.M(errors.MetaKeyAppName, a.id))
+		}
+	}()
+
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+
+	if a.srvState == nil {
+		return errors.New(ErrAppNotInstalled)
+	}
+
+	cfgStr, err := yaml.Marshal(config)
+	if err != nil {
+		return errors.New(err, errors.TypeApplication)
+	}
+	err = a.srvState.UpdateConfig(string(cfgStr))
+	if err != nil {
+		return errors.New(err, errors.TypeApplication)
+	}
+	return nil
+}
+
+// Stop stops the current application.
+func (a *Application) Stop() {
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+
+	if a.srvState == nil {
+		return
+	}
+
+	if err := a.srvState.Stop(a.processConfig.StopTimeout); err != nil {
+		a.setState(state.Failed, errors.New(err, "Failed to stopped").Error())
+	} else {
+		a.setState(state.Stopped, "Stopped")
+	}
+	a.srvState = nil
+
+	a.cleanUp()
+	a.stopCredsListener()
+}
+
+// OnStatusChange is the handler called by the GRPC server code.
+//
+// It updates the status of the application and handles restarting the application is needed.
+func (a *Application) OnStatusChange(s *server.ApplicationState, status proto.StateObserved_Status, msg string) {
+	a.appLock.Lock()
```

Comment:
same thing with unlock

Reply:
updated to use `defer a.appLock.Unlock()`